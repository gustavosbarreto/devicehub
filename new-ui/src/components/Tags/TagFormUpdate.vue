<template>
  <v-list-item v-bind="$props, $attrs" @click="showDialog = true">
    <div class="d-flex align-center">
      <div class="mr-2">
        <v-icon color="white"> mdi-tag </v-icon>
      </div>

      <v-list-item-title data-test="mdi-information-list-item">
        {{ hasTags ? "Edit tags" : "Add Tags" }}
      </v-list-item-title>
    </div>
  </v-list-item>

  <v-dialog v-model="showDialog" min-width="280" max-width="450">
    <v-card class="bg-v-theme-surface">
      <v-card-title class="text-h5 pa-4 bg-primary">
        {{ hasTags ? "Edit tags" : "Add Tags" }}
      </v-card-title>
      <v-divider />

      <v-card-text class="mt-5 w-100">
        <v-combobox
          id="targetInput"
          full-width
          ref="tags"
          v-model="inputTags"
          :error-messages="tagsError"
          label="Tag"
          hint="Maximum of 3 tags"
          multiple
          chips
          variant="outlined"
          data-test="deviceTag-combobox"
          :deletable-chips="true"
          :delimiters="[',', ' ']"
        />
      </v-card-text>

      <v-card-actions>
        <v-spacer />

        <v-btn
          variant="text"
          data-test="close-btn"
          @click="close()"
          class="mr-2"
        >
          Close
        </v-btn>

        <v-btn variant="text" data-test="save-btn" @click="save()">
          Save
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, watch } from "vue";
import { useField } from "vee-validate";
import * as yup from "yup";
import { useStore } from "../../store";
import { AxiosError } from "axios";
import {
  INotificationsError,
  INotificationsSuccess,
} from "../../interfaces/INotifications";

export default defineComponent({
  props: {
    deviceUid: {
      type: String,
      required: true,
    },

    tagsList: {
      type: Array<string>,
      required: true,
    },
  },
  emits: ["update"],
  setup(props, ctx) {
    const store = useStore();
    const showDialog = ref(false);

    const hasTags = computed(() => props.tagsList.length > 0);
    const inputTags = ref(props.tagsList);
    const tagsError = ref("");

    watch(inputTags, () => {
      if (inputTags.value.length > 3) {
        tagsError.value = "Maximum of 3 tags";
      } else {
        tagsError.value = "";
      }
    });

    const save = async () => {
      try {
        tagsError.value = "";
        await store.dispatch("devices/updateDeviceTag", {
          uid: props.deviceUid,
          tags: { tags: inputTags.value },
        });

        await store.dispatch("tags/setTags", {
          data: inputTags.value,
          headers: {
            "x-total-count": inputTags.value.length,
          },
        });
        showDialog.value = false;
        store.dispatch(
          "snackbar/showSnackbarSuccessAction",
          INotificationsSuccess.deviceTagUpdate
        );

        ctx.emit("update");
      } catch (error: any) {
        switch (error.response.status) {
          // when the name the format is invalid.
          case 400: {
            tagsError.value =
              "The format is invalid. Min 3, Max 255 characters!";
            break;
          }
          // when the user is not authorized.
          case 403: {
            store.dispatch(
              "snackbar/showSnackbarErrorAction",
              INotificationsError.deviceTagUpdate
            );
            break;
          }
          // When the array tag size reached the max capacity.
          case 406: {
            tagsError.value = "The maximum capacity has reached.";
            break;
          }
          default: {
            store.dispatch(
              "snackbar/showSnackbarErrorAction",
              INotificationsError.deviceTagUpdate
            );
          }
        }
      }
      return false;
    };

    const close = () => {
      showDialog.value = false;
      inputTags.value = props.tagsList;
    };

    return {
      inputTags,
      tagsError,
      showDialog,
      hasTags,
      save,
      close,
    };
  },
});
</script>
