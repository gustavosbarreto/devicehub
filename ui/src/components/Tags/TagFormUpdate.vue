<template>
  <v-list-item
    v-bind="$attrs"
    @click="showDialog = true"
    :disabled="notHasAuthorization"
  >
    <div class="d-flex align-center">
      <div class="mr-2">
        <v-icon> mdi-tag </v-icon>
      </div>

      <v-list-item-title data-test="mdi-information-list-item">
        {{ hasTags ? "Edit tags" : "Add Tags" }}
      </v-list-item-title>
    </div>
  </v-list-item>

  <v-dialog v-model="showDialog" min-width="280" max-width="450">
    <v-card class="bg-v-theme-surface">
      <v-card-title class="text-h5 pa-4 bg-primary">
        {{ hasTags ? "Edit tags" : "Add Tags" }}
      </v-card-title>
      <v-divider />

      <v-card-text class="mt-5 w-100">
        <v-combobox
          id="targetInput"
          full-width
          ref="tags"
          v-model="inputTags"
          :error-messages="tagsError"
          label="Tag"
          hint="Maximum of 3 tags"
          multiple
          chips
          variant="outlined"
          data-test="deviceTag-combobox"
          :deletable-chips="true"
          :delimiters="[',', ' ']"
        />
      </v-card-text>

      <v-card-actions>
        <v-spacer />

        <v-btn
          variant="text"
          data-test="close-btn"
          @click="close()"
          class="mr-2"
        >
          Close
        </v-btn>

        <v-btn variant="text" data-test="save-btn" @click="save()">
          Save
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { computed, defineComponent, ref, watch } from "vue";
import { useStore } from "../../store";
import {
  INotificationsError,
  INotificationsSuccess,
} from "../../interfaces/INotifications";

export default defineComponent({
  props: {
    deviceUid: {
      type: String,
      required: true,
    },

    tagsList: {
      type: Array<string>,
      required: true,
    },
    notHasAuthorization: {
      type: Boolean,
      default: false,
    },
  },
  emits: ["update"],
  inheritAttrs: true,
  setup(props, ctx) {
    const store = useStore();
    const showDialog = ref(false);

    const hasTags = computed(() => props.tagsList.length > 0);
    const inputTags = ref(props.tagsList);
    const tagsError = ref("");

    const tagsHasLessThan3Characters = computed(() => inputTags.value.some((tag) => tag.length < 3));

    watch(inputTags, () => {
      if (inputTags.value.length > 3) {
        tagsError.value = "Maximum of 3 tags";
      } else if (tagsHasLessThan3Characters.value) {
        tagsError.value = "The minimum length is 3 characters";
      } else {
        tagsError.value = "";
      }
    });

    const save = async () => {
      if (tagsError.value) return;
      try {
        tagsError.value = "";
        await store.dispatch("devices/updateDeviceTag", {
          uid: props.deviceUid,
          tags: { tags: inputTags.value },
        });

        await store.dispatch("tags/setTags", {
          data: inputTags.value,
          headers: {
            "x-total-count": inputTags.value.length,
          },
        });
        showDialog.value = false;
        store.dispatch(
          "snackbar/showSnackbarSuccessAction",
          INotificationsSuccess.deviceTagUpdate,
        );

        ctx.emit("update");
      } catch (error: any) {
        switch (error.response.status) {
          // when the name the format is invalid.
          case 400: {
            tagsError.value = "The format is invalid. Min 3, Max 255 characters!";
            break;
          }
          // when the user is not authorized.
          case 403: {
            store.dispatch(
              "snackbar/showSnackbarErrorAction",
              INotificationsError.deviceTagUpdate,
            );
            break;
          }
          // When the array tag size reached the max capacity.
          case 406: {
            tagsError.value = "The maximum capacity has reached.";
            break;
          }
          default: {
            store.dispatch(
              "snackbar/showSnackbarErrorAction",
              INotificationsError.deviceTagUpdate,
            );
            throw new Error(error);
          }
        }
      }
    };

    const close = () => {
      showDialog.value = false;
      inputTags.value = props.tagsList;
    };

    return {
      inputTags,
      tagsError,
      showDialog,
      hasTags,
      tagsHasLessThan3Characters,
      save,
      close,
    };
  },
});
</script>
