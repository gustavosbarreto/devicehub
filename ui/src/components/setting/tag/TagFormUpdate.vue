<template>
  <fragment>
    <v-tooltip
      :disabled="hasAuthorization"
      bottom
    >
      <template #activator="{ on }">
        <span v-on="on">
          <v-list-item-title
            data-test="title-item"
            v-on="on"
          >
            {{ hasTag ? 'Edit tag' : 'Add tag' }}
          </v-list-item-title>
        </span>

        <span v-on="on">
          <v-icon
            :disabled="!hasAuthorization"
            left
            data-test="tag-icon"
            v-on="on"
            v-text="'mdi-tag'"
          />
        </span>
      </template>

      <span v-if="!hasAuthorization">
        You don't have this kind of authorization.
      </span>
    </v-tooltip>

    <v-dialog
      v-model="showDialog"
      max-width="400"
      @click:outside="close"
    >
      <v-card data-test="tagForm-card">
        <v-card-title
          class="headline primary text-center"
          v-text="'Manage Tags'"
        />

        <v-card-text>
          <v-combobox
            id="targetInput"
            v-model="listTagLocal"
            label="Tag"
            hint="Maximum of 3 tags"
            multiple
            chips
            append-icon
            data-test="deviceTag-combobox"
            :deletable-chips="true"
            :rules="[tagRule]"
            :delimiters="[',', ' ']"
          />
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="close-btn"
            @click="close"
            v-text="'Close'"
          />

          <v-btn
            text
            data-test="save-btn"
            @click="save()"
            v-text="'Save'"
          />
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import hasPermission from '@/components/filter/permission';

export default {
  name: 'TagFormDialogComponent',

  filters: { hasPermission },

  props: {
    deviceUid: {
      type: String,
      required: true,
    },

    tagsList: {
      type: Array,
      required: true,
    },

    show: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      listTagLocal: [],
      errorMsg: '',
      action: 'deviceUpdate',
    };
  },

  computed: {
    showDialog: {
      get() {
        return this.show && this.hasAuthorization;
      },
      set(value) {
        this.$emit('update:show', value);
      },
    },

    hasTag() {
      return this.tagsList.length > 0;
    },

    hasAuthorization() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.tag[this.action],
        );
      }

      return false;
    },
  },

  watch: {
    showDialog(value) {
      if (value) {
        this.setLocalVariable();
      }
    },

    listTagLocal(list) {
      if (list.length > 3) {
        this.$nextTick(() => this.listTagLocal.pop());
        this.errorMsg = 'The maximum capacity has reached.';
      }
    },
  },

  async created() {
    await this.setLocalVariable();
  },

  methods: {
    tagRule() {
      if (this.errorMsg !== '') {
        return this.errorMsg;
      }

      return true;
    },

    setLocalVariable() {
      this.listTagLocal = this.tagsList;
      this.errorMsg = '';
    },

    async save() {
      const data = { uid: this.deviceUid, tags: this.listTagLocal };

      try {
        this.errorMsg = '';
        await this.$store.dispatch('devices/updateDeviceTag', data);
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.deviceTagUpdate);

        this.update();
      } catch (error) {
        this.$nextTick(() => this.listTagLocal.pop());
        switch (true) {
        // when the name the format is invalid.
        case (error.response.status === 400): {
          this.errorMsg = 'The format is invalid. Min 3, Max 255 characters!';
          break;
        }
        // when the user is not authorized.
        case (error.response.status === 403): {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceTagUpdate);
          break;
        }
        // When the array tag size reached the max capacity.
        case (error.response.status === 406): {
          this.errorMsg = 'The maximum capacity has reached.';
          break;
        }
        default: {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceTagUpdate);
        }
        }
      }
      return false;
    },

    update() {
      this.$emit('update');
      this.close();
    },

    close() {
      this.$emit('update:show', false);
    },
  },
};
</script>
