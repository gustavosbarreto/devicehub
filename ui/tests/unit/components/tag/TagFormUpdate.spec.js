import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import TagFormUpdate from '@/components/tag/TagFormUpdate';

describe('TagFormUpdate', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const tests = [
    {
      description: 'Dialog closed with add',
      props: {
        deviceUid: '',
        tagsList: [],
        show: false,
      },
      data: {
        dialog: false,
        listTagLocal: [],
        errorMsg: '',
      },
      computed: {
        hasTag: false,
      },
      template: {
        'edit-icon': true,
        'edit-title': true,
        'tagForm-card': false,
      },
      templateText: {
        'edit-title': 'Add tags',
      },
    },
    {
      description: 'Dialog closed with edit',
      props: {
        deviceUid: 'xxxxxxx',
        tagsList: ['ShellHub'],
        show: false,
      },
      data: {
        dialog: false,
        listTagLocal: ['ShellHub'],
        errorMsg: '',
      },
      computed: {
        hasTag: true,
      },
      template: {
        'edit-icon': true,
        'edit-title': true,
        'tagForm-card': false,
      },
      templateText: {
        'edit-title': 'Edit tags',
      },
    },
    {
      description: 'Dialog opened',
      props: {
        deviceUid: 'xxxxxxx',
        tagsList: ['ShellHub'],
        show: true,
      },
      data: {
        dialog: false,
        listTagLocal: ['ShellHub'],
        errorMsg: '',
      },
      computed: {
        hasTag: true,
      },
      template: {
        'edit-icon': true,
        'edit-title': true,
        'tagForm-card': true,
      },
      templateText: {
        'edit-title': 'Edit tags',
      },
    },
  ];

  const storeVuex = () => new Vuex.Store({
    namespaced: true,
    state: { },
    getters: { },
    actions: {
      'devices/updateDeviceTag': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
    },
  });

  tests.forEach((test) => {
    describe(`${test.description}`, () => {
      beforeEach(async () => {
        wrapper = mount(TagFormUpdate, {
          store: storeVuex(),
          localVue,
          stubs: ['fragment'],
          propsData: {
            deviceUid: test.props.deviceUid,
            tagsList: test.props.tagsList,
            show: test.props.show,
          },
          vuetify,
        });
      });

      ///////
      // Component Rendering
      //////

      it('Is a Vue instance', () => {
        expect(wrapper).toBeTruthy();
      });
      it('Renders the component', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      ///////
      // Data checking
      //////

      it('Receive data in props', () => {
        Object.keys(test.props).forEach((item) => {
          expect(wrapper.vm[item]).toEqual(test.props[item]);
        });
      });
      it('Compare data with default value', () => {
        Object.keys(test.data).forEach((item) => {
          expect(wrapper.vm[item]).toEqual(test.data[item]);
        });
      });
      it('Process data in the computed', () => {
        Object.keys(test.computed).forEach((item) => {
          expect(wrapper.vm[item]).toEqual(test.computed[item]);
        });
      });

      //////
      // HTML validation
      //////

      it('Renders the template with data', () => {
        Object.keys(test.template).forEach((item) => {
          expect(wrapper.find(`[data-test="${item}"]`).exists()).toBe(test.template[item]);
        });
      });
      it('Renders template with expected text', () => {
        Object.keys(test.templateText).forEach((item) => {
          expect(wrapper.find(`[data-test="${item}"]`).text()).toContain(test.templateText[item]);
        });
      });
    });
  });
});
