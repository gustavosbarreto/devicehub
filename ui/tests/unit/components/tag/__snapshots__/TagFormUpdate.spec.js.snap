// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagFormUpdate Dialog closed with add Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__icon mr-0"><i aria-hidden="true" data-test="edit-icon" class="v-icon notranslate v-icon--left mdi mdi-tag theme--light"></i></div>
  <div class="v-list-item__content">
    <div data-test="edit-title" class="v-list-item__title text-left">Add tags</div>
  </div>
  <div class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`TagFormUpdate Dialog closed with edit Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__icon mr-0"><i aria-hidden="true" data-test="edit-icon" class="v-icon notranslate v-icon--left mdi mdi-tag theme--light"></i></div>
  <div class="v-list-item__content">
    <div data-test="edit-title" class="v-list-item__title text-left">Edit tags</div>
  </div>
  <div class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`TagFormUpdate Dialog opened Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__icon mr-0"><i aria-hidden="true" data-test="edit-icon" class="v-icon notranslate v-icon--left mdi mdi-tag theme--light"></i></div>
  <div class="v-list-item__content">
    <div data-test="edit-title" class="v-list-item__title text-left">Edit tags</div>
  </div>
  <div class="v-dialog__container"></div>
</fragment-stub>
`;
